src/adaptors/tcp/tcp_adaptor.c

```diff
@@ -1936,6 +1936,28 @@ static int setup_tls_session(qd_tcp_connection_t *conn, const qd_tls_domain_t *p
     return 0;
 }
 
+/**
+ * Initiate the termination of a group of connections.
+ *
+ * @param connections Reference to the list of connections
+ */
+static void terminate_connections(qd_tcp_connection_list_t *connections)
```

Comment:
There is a function called `close_connection_XSIDE_IO()` which can be used to terminate tcp connections. It take care of freeing `qd_tcp_connection_t` objects and well as its corresponding `qdr_connection_t` objects
Like so - 
```
static void terminate_listener_connections(qd_tcp_listener_t *listener) {
        //
        // Deliberately call qd_tcp_listener_incref() to make sure that freeing the connections, don't
        // free the listener. Then we call qd_tcp_listener_free() to forcefully free the listener without checking the listener->ref_count
        //
        qd_tcp_listener_incref(listener);
        qd_tcp_connection_t *conn = DEQ_HEAD(listener->connections);
        while (conn) {
            qd_tcp_connection_t *next_conn = DEQ_NEXT(conn);
            close_connection_XSIDE_IO(conn);
            conn = next_conn;
        }
        qd_tcp_listener_free(listener);
}
```

Reply:
I thought that `close_connection_XSIDE_IO()` cannot be called directly in a loop because it makes proton calls on the connection. Another proactor thread may process the same connection at the same time which is prohibited. I maybe misunderstood this paragraph from `threading-info.txt`: https://github.com/skupperproject/skupper-router/blob/35d5a0fae00af782663e95f4ebea69299bb7edd6/docs/notes/threading-info.txt#L50

Reply:
No, you did not misunderstand the paragraph. What I suggested - calling close_connection_XSIDE_IO() - from a management thread - is wrong. I will get back to you shortly

Reply:
You might try calling `qdr_core_close_connection(conn->core_conn)`
like here - https://github.com/skupperproject/skupper-router/blob/main/src/adaptors/http2/http2_adaptor.c#L3403

Reply:
Okay, thanks for the pointer! If I call that function then I do not need to change the type of the `closed` flag from `bool` to `sys_atomic_t` in `qdr_connection_t`. Is there any other reason why the current PR is not a proper solution? I'm just asking because I would like to understand the router code better :-).

Reply:
Yes, if you call `qdr_core_close_connection(conn->core_conn)`, you will not need to change the type on the closed flag to `sys_atomic_t` in `qdr_connection_t`. 

Notice that the closed flag is accessed in two places, inside the core thread callbacks and in the `qdr_connection_process()` under a `conn->work_lock`.
The core thread is single threaded and so most of the core thread data structures don't need locking when accessed by the core thread except during connection activation which happens on an I/O thread. 

We usually tend to have I/O thread functions which then employ the core thread to modify these data structures via core thread callbacks. So, we generally access `conn->closed` from a core thread callback and `qdr_core_close_connection(conn->core_conn)` does that for you.

Reply:
Two things
1. You don't need to call `pn_raw_connection_wake()` since it is already called as part of the call to `qdr_core_close_connection()`
2. You also don't need to lock on the activaton_lock since you are not going to call `pn_raw_connection_wake()` anymore

Reply:
Just to make sure that I understand correctly the code, would the call sequence be roughly be as follows:

1.  `qdr_core_close_connection() ` will activate the core thread
2.  qdr_core_close_connection_CT() will set `conn->closed=true` and activate the connection
3. the TCP adaptor activation callback `CORE_activate()` will call `pn_raw_connection_wake()`

I think the check for `conn->raw_opened` can also be removed from  `terminate_connections()` because the same check is done in `CORE_active()`.

Reply:
1. `qdr_core_close_connection(qdr_connection_t *conn)` will add a core thread action with callback `qdr_core_close_connection_CT()` and wake up the core thread 
2. The `qdr_core_close_connection_CT()` is executed by the core thread and will set the  `conn->closed=true` and activate the connection to be deleted via a call to `qdr_connection_activate_CT(core,  conn);` 
3. `qdr_connection_process()' in `connections.c` is executed on an I/O thread and an adaptor callback function is called (line 359 of connections.c) - `conn->protocol_adaptor->conn_close_handler(conn->protocol_adaptor->user_context, conn, conn->error);`
4. The `conn_close_handler` is set here - https://github.com/skupperproject/skupper-router/blob/main/src/adaptors/tcp/tcp_adaptor.c#L2548 which is `static void CORE_connection_close(void *context, qdr_connection_t *conn, qdr_error_t *error)` where there is a call to `pn_raw_connection_close(tcp_conn->raw_conn);` which will generate a PN_RAW_CONNECTION_DISCONNECTED proton event and the handler for that event ultimately calls `close_connection_XSIDE_IO(conn)` where the `qd_tcp_connection_t` object and its contents are ultimately freed.

Reply:
Thanks!